notebooks/02_model/vowpal_wabbit_deep_dive.ipynb

```diff
@@ -0,0 +1,737 @@
+    "[Vowpal Wabbit](https://github.com/VowpalWabbit/vowpal_wabbit) is a fast online machine learning library that implements several algorithms relevant to the recommendation use case.\n",
+    "\n",
+    "The main advantage of Vowpal Wabbit (VW) is that training is done in an online fashion typically using Stochastic Gradient Descent or similar variants, which allows it to scale well to very large datasets. Additionally, it is optimized to run very quickly and can support distributed training scenarios for extremely large datasets.\n",
+    "\n",
+    "In this notebook we demonstrate how to use the VW library to generate recommendations on the [Movielens](https://grouplens.org/datasets/movielens/) dataset.\n",
+    "\n",
+    "Several things are worth noting in how VW is being used in this notebook. By leveraging an Azure Data Science Virtual Machine ([DSVM](https://azure.microsoft.com/en-us/services/virtual-machines/data-science-virtual-machines/)), VW comes pre-installed and can be used directly from the command line. There are also python bindings to be able to use VW within a python environment and even a wrapper conforming to the SciKit-Learn Estimator API. However, the python bindings must be installed as an additional python package with Boost dependencies, so for simplicity's sake execution of VW is done via a subprocess call mimicking what would happen from the command line execution of the model.\n",
+    "\n",
+    "VW expects a specific [input format](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format), and to_vw() below is a convenience function to convert the standard movielens dataset into that format. Datafiles then are written to disk and passed to VW for training.\n",
+    "\n",
+    "The examples shown are to demonstrate functional capabilities not to indicate performance advantages of different approaches. There are several hyper-parameters that can greatly impact performance of VW models (see [command line options](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Command-Line-Arguments)). To properly compare approaches it is helpful to learn about and tune these parameters for production workloads."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<h3>Environment Setup</h3>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "from subprocess import run\n",
+    "from tempfile import TemporaryDirectory\n",
+    "\n",
+    "import pandas as pd\n",
+    "\n",
+    "from reco_utils.dataset.movielens import load_pandas_df\n",
+    "from reco_utils.dataset.python_splitters import python_stratified_split\n",
+    "from reco_utils.evaluation.python_evaluation import rmse, mae, exp_var, rsquared, get_top_k_items"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def to_vw(df, output, logistic=False):\n",
```

Comment:
not sure if we are going to reuse this function. If we are going to do several notebooks about VW, maybe it would be worthwhile to move this file to a new vw folder under `reco_utils/recommender`

Reply:
imo this is a bit of a workaround that shouldn't be pushed forward for general use

Reply:
that said it might be worth considering this relative to output format creation for various libraries, if this is something that would be used to convert a dataframe or pipe a file through it could be generalized.

Reply:
makes sense